File: 卡码/代码随想录/1/05-栈与队列/Go/03-逆波兰表达式/LeetCode_0150.go
// 150. 逆波兰表达式求值
package main

import "strconv"

func evalRPN(tokens []string) int {
	nums:=make([]int,0)
	var n1,n2 int
	for _,token:=range tokens {
		if token=="+"||token=="-"||token=="*"||token=="/" {
			n2=nums[len(nums)-1]
			n1=nums[len(nums)-2]
			nums=nums[0:len(nums)-2]
		}
		switch token {
			case "+":
				nums=append(nums,n1+n2)
			case "-":
				nums=append(nums,n1-n2)
			case "*":
				nums=append(nums,n1*n2)
			case "/":
				nums=append(nums,n1/n2)
			default:
				num,_:=strconv.Atoi(token)
				nums=append(nums,num)
		}
	}
	return nums[len(nums)-1]
}

File: 卡码/代码随想录/Python/栈与队列/stackAndQueue/03-逆波兰表达式/LeetCode_0150.py
#150. 逆波兰表达式求值
from typing import *
import math

class Solution:
  def evalRPN(self, tokens: List[str]) -> int:
    nums:List[int]=[]
    for token in tokens:
      if token=="+" or token=="-" or token=="*" or token=="/":
        n2=nums.pop()
        n1=nums.pop()
        if token=="+":
          nums.append(n1+n2)
        elif token=="-":
          nums.append(n1-n2)
        elif token=="*":
          nums.append(n1*n2)
        else:
          num=n1/n2
          nums.append(math.floor(num) if num>0 else math.ceil(num))
      else:
        nums.append(int(token))
    return nums[-1]